### PR TITLE
Fixes races around {api.Module, wazero.Runtime}.CloseWithExitCode

### PR DESCRIPTION
**Pull request by mathetake**
This is a regression introduced in #1108 and hasn't been well covered in the unit tests with -race.
Thanks to @ckaznocha for reporting this!

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: internal/wasm/call_context.go [hunk 1]
```diff
@@ -122,14 +124,24 @@ func (m *CallContext) Close(ctx context.Context) (err error) {
...
-func (m *CallContext) setExitCode(exitCode uint32) {
+// closeWithExitCode is the same as CloseWithExitCode besides this doesn't delete it from Store.moduleList.
+func (m *CallContext) closeWithExitCode(ctx context.Context, exitCode uint32) (err error) {
+	if !m.setExitCode(exitCode) {
+		return nil
+	}
+	return m.ensureResourcesClosed(ctx)
+}
+
+func (m *CallContext) setExitCode(exitCode uint32) bool {
```
Comment (mathetake) [inline internal/wasm/call_context.go:142]: I accidentally deleted the boolean of the `swap` from `CompareAndSwapUint64` and dropped the check. so fixed it!

File: internal/wasm/call_context.go [hunk 2]
```diff
@@ -122,14 +124,24 @@ func (m *CallContext) Close(ctx context.Context) (err error) {
 
 // CloseWithExitCode implements the same method as documented on api.Module.
 func (m *CallContext) CloseWithExitCode(ctx context.Context, exitCode uint32) (err error) {
-	m.setExitCode(exitCode)
+	if !m.setExitCode(exitCode) {
+		return nil
```
Comment (codefromthecrypt) [inline internal/wasm/call_context.go:128]:
```suggestion
		return nil // not an error to have already closed
```

File: internal/wasm/call_context.go [hunk 3]
```diff
@@ -122,14 +124,24 @@ func (m *CallContext) Close(ctx context.Context) (err error) {
...
+	}
 	_ = m.s.deleteModule(m.Name())
 	return m.ensureResourcesClosed(ctx)
 }
 
-func (m *CallContext) setExitCode(exitCode uint32) {
+// closeWithExitCode is the same as CloseWithExitCode besides this doesn't delete it from Store.moduleList.
+func (m *CallContext) closeWithExitCode(ctx context.Context, exitCode uint32) (err error) {
+	if !m.setExitCode(exitCode) {
+		return nil
```
Comment (codefromthecrypt) [inline internal/wasm/call_context.go:137]:
```suggestion
		return nil // not an error to have already closed
```

File: internal/wasm/call_context_test.go [hunk 4]
```diff
@@ -147,30 +148,46 @@ func TestCallContext_Close(t *testing.T) {
...
-		// We use side effects to determine if Close in fact called Context.Close (without repeating sys_test.go).
-		// One side effect of Context.Close is that it clears the openedFiles map. Verify our base case.
-		_, ok := fsCtx.LookupFile(3)
-		require.True(t, ok, "sysCtx.openedFiles was empty")
-
-		// Closing should not err.
-		require.NoError(t, m.Close(testCtx))
+				// Closing should not err.
+				if concurrent {
+					hammer.NewHammer(t, 100, 10).Run(func(name string) {
```
Comment (codefromthecrypt) [inline internal/wasm/call_context_test.go:171]: nit: indentation is getting intense, but ok. if you see a cheap way to reduce this, by extracting a new top-level test, or a function do it.

**Review by codefromthecrypt — APPROVED** [review]
thanks for fixing this!

**Pull request merged**